### PR TITLE
jungleyo.cpp: fix video emulation, currently sorta playable [Angelo Salese]

### PR DESCRIPTION
**Pull request by angelosa**
Several gameplay issues should be down to imperfect decryption.

### CODE REVIEW

**Comment by angelosa** [comment]
@DavidHaywood @Osso13 please confirm about not being credited in copyright holder.

**Comment by DavidHaywood** [comment]
Anything I did was a skeleton, and I don't even remember that, so no, I don't require any attribution here.

**Comment by DavidHaywood** [comment]
on a similar note, is there anything you'd consider to be your code left in xavix.cpp? you're still credited there

**Comment by angelosa** [comment]
> 
> 
> on a similar note, is there anything you'd consider to be your code left in xavix.cpp? you're still credited there

Nothing worth for me being credited, was a skeleton with wrong CPU type back in MESS days wow.

**Comment by Osso13** [comment]
No need to credit me, thanks.

File: src/mame/drivers/jungleyo.cpp [hunk 1]
```diff
@@ -90,86 +112,179 @@ class jungleyo_state : public driver_device
 	uint32_t screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
 	TILE_GET_INFO_MEMBER(get_bg_tile_info);
 	TILE_GET_INFO_MEMBER(get_fg_tile_info);
-	void bg_videoram_w(offs_t offset, uint16_t data);
-	void fg_videoram_w(offs_t offset, uint16_t data);
+	template <int layer_num> TILE_GET_INFO_MEMBER(get_reel_tile_info);
+	void bg_videoram_w(offs_t offset, u16 data, u16 mem_mask = ~0);
+	void fg_videoram_w(offs_t offset, u16 data, u16 mem_mask = ~0);
+	template <int layer_num> void reel_vram_w(offs_t offset, u16 data, u16 mem_mask = ~0);
```
Comment (cuavas) [inline src/mame/drivers/jungleyo.cpp:118]: Please use `LlamaCase` for template parameters to make it clear what they are.

File: src/mame/drivers/jungleyo.cpp [hunk 2]
```diff
@@ -90,86 +112,179 @@ class jungleyo_state : public driver_device
...
 {
 	m_bg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_bg_tile_info)), TILEMAP_SCAN_ROWS, 8, 8, 64, 32);
 	m_fg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_fg_tile_info)), TILEMAP_SCAN_ROWS, 8, 8, 64, 32);
+	m_reel_tilemap[0] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<0>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+	m_reel_tilemap[1] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<1>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+	m_reel_tilemap[2] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<2>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+
 	m_fg_tilemap->set_transparent_pen(0);
+	m_bg_tilemap->set_transparent_pen(0);
+	for (int i = 0;i < 3; i++)
```
Comment (cuavas) [inline src/mame/drivers/jungleyo.cpp:196]: Missing space after the first semicolon here.  (You could do this with a range-based `for` loop, too – `for (auto &reel : m_reel_tilemap) { reel->… }`).

File: src/mame/drivers/jungleyo.cpp [hunk 3]
```diff
@@ -178,21 +293,23 @@ void jungleyo_state::main_map(address_map &map)
...
-	map(0xb10000, 0xb1ffff).ram();
-	map(0xb20000, 0xb2ffff).ram();
+	map(0xb00000, 0xb2ffff).rw(FUNC(jungleyo_state::palette_ram_r), FUNC(jungleyo_state::palette_ram_w)).umask16(0x00ff);
 	map(0xb80000, 0xb81fff).ram().share(m_fg_videoram).w(FUNC(jungleyo_state::fg_videoram_w));
 	map(0xb90000, 0xb91fff).ram().share(m_bg_videoram).w(FUNC(jungleyo_state::bg_videoram_w));
-	map(0xba0000, 0xba1fff).ram();
+	map(0xba0000, 0xba07ff).ram().share(m_reel_vram[0]).w(FUNC(jungleyo_state::reel_vram_w<0>));
+	map(0xba0800, 0xba0fff).ram().share(m_reel_vram[1]).w(FUNC(jungleyo_state::reel_vram_w<1>));
+	map(0xba1000, 0xba17ff).ram().share(m_reel_vram[2]).w(FUNC(jungleyo_state::reel_vram_w<2>));
+	map(0xba1800, 0xba1fff).ram(); // supposedly the 4th reel VRAM bank, inited at POST only and prolly just tied to NOP instead
```
Comment (cuavas) [inline src/mame/drivers/jungleyo.cpp:311]: Is it really too much to spell “probably” correctly?

Comment (angelosa) [reply src/mame/drivers/jungleyo.cpp:311]: Excuse me?

cuavas_prolly

Comment (DavidHaywood) [reply src/mame/drivers/jungleyo.cpp:311]: it is very informal (and I can't remember the last time anybody I know here used it)

Comment (Tafoid) [reply src/mame/drivers/jungleyo.cpp:311]: I used to use it a lot for online games back in the 80's/90's/00's .. but have not used it all that much in recent times.

Comment (cuavas) [reply src/mame/drivers/jungleyo.cpp:311]: It’s not a real word, it’s a phonetic spelling of a lazy mispronunciation.  We have higher standards than that.

Comment (angelosa) [reply src/mame/drivers/jungleyo.cpp:311]: 7 hits in the source code, and this is the first time I hear that "prolly" is against "standards" (?).

File: src/mame/drivers/jungleyo.cpp (hunk 2 above)
Comment (angelosa) [reply src/mame/drivers/jungleyo.cpp]: For some reason I forgot that C++ has a _for in_ equivalent, fixed.
I haven't tried any of [these](https://stackoverflow.com/questions/841075/best-c-code-formatter-beautifier) suggestions, any good one that can be also a good candidate for _srcclean_ phase?

File: src/mame/drivers/jungleyo.cpp (hunk 3 above)
Comment (einstein95) [reply src/mame/drivers/jungleyo.cpp]: The 6 cases in the MAME code, organised chronologically:
28 Jun 2017: https://github.com/mamedev/mame/commit/12824956e936e9716a7d29c9dd18f12aeec9632d#diff-b98cf4a01eb48be4ccf23fdefa7c84f2a8aa300e595e3ea4abf3a7e7cf85050aR40
15 Feb 2018: https://github.com/mamedev/mame/commit/852a6533e9f2afd3f4d7589a7f254b782a54b610#diff-56bb0d2e3fba56f29dfe5497594d12ed60a3d56fddb6047f5726e84c830c8dffR94
24 May 2019: https://github.com/mamedev/mame/commit/9af3931f194613761eae93a20d7bf8495031e2f4#diff-7fff0e3c1abc3bed817c80ae0909b4cff0623eaddd7842d0f09e8b09805ad2bfR518
24 Jul 2019: https://github.com/mamedev/mame/commit/e721e1f4ff4572876bebee1da56d8b6f27745e2c#diff-a9f92a3049a8a9c19ad7a5b99ecadba8a4d38864ecea1a249dcdc2c97cddc50fR538
24 Aug 2019: https://github.com/mamedev/mame/commit/f5b2fd9ee6ef6ced1eec7859494e82c375d9c8d2#diff-148eb445a116bce4716f4f53f037d4453755ab463cb8f32830057df3d73d3197R243
13 Dec 2019: https://github.com/mamedev/mame/commit/f58c3ea2644c51559a45d09fa1c3a146410bcf5a#diff-64372177e8768593bf91f048ae7ec7e493858326cbf183269f60b361e641faaaR734

**Pull request merged**